x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -254,6 +237,50 @@ public void setEnabled(String uid, boolean enabled, RefreshPolicy refreshPolicy,
         doUpdate(buildUpdateRequest(uid, builder, refreshPolicy, -1, -1), listener.map(updateResponse -> AcknowledgedResponse.TRUE));
     }
 
+    // package private for testing
+    SearchRequest buildSearchRequest(SuggestProfilesRequest request) {
+        final BoolQueryBuilder query = QueryBuilders.boolQuery().filter(QueryBuilders.termQuery("user_profile.enabled", true));
+        if (Strings.hasText(request.getName())) {
+            query.must(
+                QueryBuilders.multiMatchQuery(
+                    request.getName(),
+                    "user_profile.user.username",
+                    "user_profile.user.username._2gram",
+                    "user_profile.user.username._3gram",
+                    "user_profile.user.full_name",
+                    "user_profile.user.full_name._2gram",
+                    "user_profile.user.full_name._3gram",
+                    "user_profile.user.email"
+                ).type(MultiMatchQueryBuilder.Type.BOOL_PREFIX).fuzziness(Fuzziness.AUTO)
+            );
+        }
+        final SuggestProfilesRequest.Hint hint = request.getHint();
+        if (hint != null) {
+            final List<String> hintedUids = hint.getUids();
+            if (hintedUids != null) {
+                assert false == hintedUids.isEmpty() : "uids hint cannot be empty";
+                query.should(QueryBuilders.termsQuery("user_profile.uid", hintedUids));
+            }
+            final Tuple<String, List<String>> label = hint.getSingleLabel();
+            if (label != null) {
+                final List<String> labelValues = label.v2();
+                if (labelValues.size() == 1) {
+                    query.should(QueryBuilders.termQuery("user_profile.labels." + label.v1(), labelValues.get(0)));
+                } else {
+                    query.should(QueryBuilders.termsQuery("user_profile.labels." + label.v1(), labelValues));
+                }
```

Comment:
I wonder why the term/terms distinction for "labels" but not for "uid".

Reply:
In short, I forgot. But on a second thought, the distincation does not seem necessary. I remember seeing this pattern in other places and I assumed it is for performance. But I cannot find it in other places now and there are examples where "terms" is used for even for a single value. So I updated to always use termsQuery.